tapdb/multiverse_cache.go

```diff
@@ -187,6 +189,212 @@ func (r *rootIndex) wipe() {
 	r.Store(&idx)
 }
 
+// syncerRootNodeCache is used to cache the set of active root nodes for the
+// multiverse tree, which is specifically kept for the universe sync.
+type syncerRootNodeCache struct {
+	sync.RWMutex
+
+	// universeKeyList is the list of all keys of the universes that are
+	// currently known in this multiverse. This slice is sorted by the key
+	// (which might differ from the database ordering) and can be used to
+	// paginate through the universes efficiently. The universe identifier
+	// key also contains the proof type, so there will be two entries
+	// per asset ID or group key, one for issuance and one for transfer
+	// universes. The universe roots map below will contain a root for each
+	// of these universes.
+	universeKeyList []universe.IdentifierKey
+
+	// universeRoots is a map of universe ID key to the root of the
+	// universe. This map is needed to look up the root of an individual
+	// universe quickly, or to look up the roots of all the universes when
+	// paging through them. This map never needs to be cleared, as universe
+	// roots are only added (new issuance or transfer events) or modified
+	// (re-issuance into grouped asset or new transfer) but never removed.
+	// Therefore, when the cache needs to be wiped, we only need to clear
+	// and re-fetch the universe list above to make sure new universes are
+	// included. Then we can just overwrite the universe roots map with the
+	// new roots, but skip static roots (e.g. non-grouped issuance roots).
+	universeRoots map[universe.IdentifierKey]universe.Root
+
+	*cacheLogger
+}
+
+// newSyncerRootNodeCache creates a new root node cache.
+func newSyncerRootNodeCache() *syncerRootNodeCache {
+	return &syncerRootNodeCache{
+		universeRoots: make(
+			map[universe.IdentifierKey]universe.Root,
+			numCachedProofs,
```

Comment:
I think we can pre-size this to a larger size, as rn it's set to 50k, but we have ~175k assets on mainnet rn.